lib/pure/pegs.nim

```diff
@@ -705,59 +743,232 @@ proc rawMatch*(s: string, p: Peg, start: int, c: var Captures): int {.
+  ##            of "Value":
+  ##              try:
+  ##                valStack.add matchStr.parseFloat
+  ##                echo valStack
+  ##              except ValueError:
+  ##                discard
+  ##            of "Sum", "Product":
+  ##              try:
+  ##                let val = matchStr.parseFloat
+  ##              except ValueError:
+  ##                if valStack.len > 1 and opStack.len > 0:
+  ##                  valStack[^2] = case opStack[^1]
+  ##                  of '+': valStack[^2] + valStack[^1]
+  ##                  of '-': valStack[^2] - valStack[^1]
+  ##                  of '*': valStack[^2] * valStack[^1]
+  ##                  else: valStack[^2] / valStack[^1]
+  ##                  valStack.setLen valStack.high
+  ##                  echo valStack 
+  ##                  opStack.setLen opStack.high
+  ##                  echo opStack 
+  ##      pkChar:
+  ##        leave:
+  ##          if length == 1 and "Value" != pStack[^1]:
+  ##            let matchChar = txt[start]
+  ##            opStack.add matchChar
+  ##            echo opStack
+  ##
+  ##  let pLen = parseArithExpr(txt)
+  ## 
+  ## The *handlers* parameter consists of code blocks for *PegKinds*,
+  ## which define the grammar elements of interest. Each block can contain
+  ## handler code to be executed when the parser enters and leaves text
+  ## matching the grammar element. An *enter* handler can access the specific
+  ## PEG AST node being matched as *p* and the position of the matched text
+  ## segment as *start*. A *leave* handler can access *p*, *start* and
+  ## also the length of the matched text segment as *length*. Symbols
+  ## declared in an *enter* handler can be made visible in the corresponding
+  ## *leave* handler by annotating them with an *inject* pragma.
+  template mkEnterCb(cbPostf, body) {.dirty.} =
+    template `enter cbPostf`(p, start) =
+      body
+
+  template mkLeaveCb(cbPostf, body) {.dirty.} =
+    template `leave cbPostf`(p, start, length) =
+      body
+
+  macro mkHandlers(hdlrs: untyped): untyped =
+    let cbms = newStmtList()
+    for co in hdlrs[0]:
+      if nnkCall != co.kind:
+        error("Call syntax expected.", co)
+      let pk = co[0]
+      if nnkIdent != pk.kind:
+        error("PegKind expected.", pk)
+      if 2 == co.len:
+        for cb in co[1]:
+          if nnkCall != cb.kind:
+            error("Call syntax expected.", cb)
+          if nnkIdent != cb[0].kind:
+            error("Handler identifier expected.", cb[0])
+          if 2 == cb.len:
+            let cbPostf = toNimIdent(substr($pk.ident, 2))
+            case $cb[0].ident
+            of "enter":
+              cbms.add getAst(mkEnterCb(cbPostf, cb[1]))
+            of "leave":
+              cbms.add getAst(mkLeaveCb(cbPostf, cb[1]))
+            else:
+              error(
+                "Unsupported handler identifier, expected 'enter' or 'leave'.",
+                cb[0]
+              )
+    cbms
+
+  block:
```

Comment:
General remarks for this entire section:
 - Some comments about what this code is doing would be nice.
    Macro code tends to be especially opaque due to the structures being operated on.
 - I feel like some of the identifiers could be given more descriptive names (`cbPostf`, `p`, `pk`, etc).

Reply:
Did that, but some variable names are determined by Araq's original code (now in *matchOrParse*).